src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,629 @@
+/** @file
+ *
+ * Marvell AQtion family network card driver.
+ *
+ * Copyright(C) 2017-2024 Marvell
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ *  TO,THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ *   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ *   OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ *   EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+FILE_LICENCE ( BSD2 );
+
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <byteswap.h>
+#include <ipxe/netdevice.h>
+#include <ipxe/ethernet.h>
+#include <ipxe/if_ether.h>
+#include <ipxe/iobuf.h>
+#include <ipxe/malloc.h>
+#include <ipxe/pci.h>
+#include <ipxe/profile.h>
+
+#include "aqc1xx.h"
+
+extern struct atl_hw_ops atl_hw;
+extern struct atl_hw_ops atl2_hw;
+
+/** @file
+*
+* Marvell AQC network card driver
+*
+*/
+
+static int atl_ring_alloc ( const struct atl_nic *nic, struct atl_ring *ring,
+			    uint32_t desc_size, uint32_t reg_base ) {
+	physaddr_t phy_addr;
+
+	/* Allocate ring buffer.*/
+	ring->length = ATL_RING_SIZE * desc_size;
+	ring->ring = dma_alloc ( nic->dma, &ring->map, ring->length,
+				ring->length );
+
+	if ( !ring->ring )
+		return -ENOMEM;
+
+	/* Initialize the descriptor ring */
+	memset ( ring->ring, 0, ring->length );
+
+	/* Program ring address */
+	phy_addr = dma ( &ring->map, ring->ring );
+
+	/* Write ring address (hi & low parts).*/
+	ATL_WRITE_REG ( ( uint32_t )phy_addr, reg_base );
+	ATL_WRITE_REG ( ( uint32_t ) ( ( ( uint64_t )phy_addr ) >> 32 ), reg_base + 4 );
```

Comment:
Maybe this should be verified to be intended?
```
/src/drivers/net/marvell/aqc1xx.c: 78 in atl_ring_alloc()
72     
73      /* Program ring address */
74      phy_addr = dma ( &ring->map, ring->ring );
75     
76      /* Write ring address (hi & low parts).*/
77      ATL_WRITE_REG ( ( uint32_t )phy_addr, reg_base );
>>>     CID 1599569:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
>>>     "(uint64_t)phy_addr >> 32" is 0 regardless of the values of its operands. This occurs as an initializer.
78      ATL_WRITE_REG ( ( uint32_t ) ( ( ( uint64_t )phy_addr ) >> 32 ), reg_base + 4 );
79     
80      /* Write ring length.*/
81      ATL_WRITE_REG ( ATL_RING_SIZE, reg_base + 8 );
82     
83      ring->sw_head = ring->sw_tail = 0;
```

Reply:
Thanks.  @animeshbm this was picked up by Coverity Scan on the nightly run.  It's a quirk of the fact that `physaddr_t` is 32-bit as far as the Coverity run is concerned.

Other drivers tend to handle this with checks on the size of `physaddr_t`, e.g. https://github.com/ipxe/ipxe/blob/804f35cb5a6ac7d35912822372b74fd8856f850b/src/drivers/bus/pcimsix.c#L190-L196

I'm not sure why it's complaining about this instance while others (e.g. in `realtek.c`) seem to be ok.  I haven't looked at the obscuring `ATL_WRITE_REG()` macro to see what it might be doing that could explain it.

@animeshbm could you please reproduce in Coverity and suggest a suitable fix?

Reply:
Sure, @mcb30 and @NiKiZe. I’ll try to reproduce the issue and let you guys know what I find.

Reply:
Thank you!